common/ontology.py

```diff
@@ -0,0 +1,56 @@
+import logging
+import random
+import time
+
+from ontoma.interface import OnToma
+from pandarallel import pandarallel
+
+ONTOMA_MAX_ATTEMPTS = 5
+pandarallel.initialize()
+
+
+def _ontoma_udf(row, ontoma_instance):
+    disease_name, disease_id = row['diseaseFromSource'], row['diseaseFromSourceId']
+    for attempt in range(1, ONTOMA_MAX_ATTEMPTS + 1):
+        # Try to map first by disease name (because that branch of OnToma is more stable), then by disease ID.
+        try:
+            mappings = ontoma_instance.find_term(query=disease_name, code=False)
+            if not mappings:
+                mappings = ontoma_instance.find_term(query=disease_id, code=True)
+            return [m.id_ot_schema for m in mappings]
+        except:
+            # If this is not the last attempt, wait until the next one
+            if attempt != ONTOMA_MAX_ATTEMPTS:
+                time.sleep(10 + 30 * random.random())
+    logging.error(f'OnToma lookup failed for {disease_name!r} / {disease_id!r}')
+    return []
+
+
+def add_efo_mapping(evidence_strings, spark_instance, ontoma_cache_dir=None):
+    """Given evidence strings with diseaseFromSource and diseaseFromSourceId fields, try to populate EFO mapping
+    field diseaseFromSourceMappedId. In case there are multiple matches, the evidence strings will be exploded
+    accordingly."""
+    logging.info('Collect all distinct (disease name, disease ID) pairs.')
+    disease_info_to_map = (
+        evidence_strings
+        .select('diseaseFromSource', 'diseaseFromSourceId')
```

Comment:
What if for certain datasources doesn't provide `diseaseFromSourceId` column? I'm afraid it would break. I can't really came up with any examples for now though.

Reply:
If all datasources have this column it might not be a blocker, just this requirement is hidden from the parser.

Reply:
That's a fair point. In the interest of time, I suggest we keep things as they are for now, because at least PhenoDigm and PanelApp do contain both fields. If we need to support datasets without `diseaseFromSourceId`, we can add them in the future.

For now, I've added a comment to make things for explicit for anyone trying to use the module: https://github.com/opentargets/evidence_datasource_parsers/pull/94/commits/101c532e48752fbae904b34d3ecfe4797ee3b5d2